File: backend/common/richtext.py
"""Sanitização de HTML permitido em campos de rich text (notícias, roadmap)."""

from __future__ import annotations

import bleach
from django.utils.html import strip_tags

ALLOWED_TAGS = [
    "p",
    "br",
    "strong",
    "b",
    "em",
    "i",
    "u",
    "s",
    "ul",
    "ol",
    "li",
    "a",
    "h2",
    "h3",
    "blockquote",
    "code",
]
ALLOWED_ATTRIBUTES = {"a": ["href", "title", "rel", "target"]}
ALLOWED_PROTOCOLS = ["http", "https", "mailto"]


def is_rich_text_empty(value: str | None) -> bool:
    """Indica se o HTML não contém texto legível após remover as tags."""
    text = strip_tags(value or "").replace("\xa0", " ").strip()
    return not text


def sanitize_rich_text(value: str | None) -> str:
    """Remove tags e atributos perigosos, preservando a formatação permitida.

    Texto puro legado (sem tags) permanece como texto. Markup oco vira ``""``.
    """
    raw = (value or "").strip()
    if not raw:
        return ""
    cleaned = bleach.clean(
        raw,
        tags=ALLOWED_TAGS,
        attributes=ALLOWED_ATTRIBUTES,
        protocols=ALLOWED_PROTOCOLS,
        strip=True,
    ).strip()
    return "" if is_rich_text_empty(cleaned) else cleaned

File: frontend/src/lib/rich-text.ts
import DOMPurify from 'dompurify'

const ALLOWED_TAGS = [
  'p', 'br', 'strong', 'b', 'em', 'i', 'u', 's', 'ul', 'ol', 'li',
  'a', 'h2', 'h3', 'blockquote', 'code',
]
const ALLOWED_ATTR = ['href', 'title', 'rel', 'target', 'class']

/** Remove tags e deixa só o texto legível. */
export function plainTextFromRichText(html: string): string {
  return (html || '')
    .replace(/<[^>]*>/g, ' ')
    .replace(/&nbsp;/gi, ' ')
    .replace(/&amp;/gi, '&')
    .replace(/&lt;/gi, '<')
    .replace(/&gt;/gi, '>')
    .replace(/\s+/g, ' ')
    .trim()
}

/** True quando não há texto útil (ex.: `<p></p>`). */
export function isRichTextEmpty(html: string): boolean {
  return !plainTextFromRichText(html)
}

function escapeHtml(text: string): string {
  return text
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
}

/** Converte texto legado ou HTML em markup seguro para exibição. */
export function sanitizeRichText(raw: string): string {
  const value = raw ?? ''
  if (!value.trim()) return ''
  const looksHtml = /<\/?[a-z][\s\S]*>/i.test(value)
  const html = looksHtml
    ? value
    : value
      .split('\n')
      .map((line) => `<p>${escapeHtml(line) || '<br>'}</p>`)
      .join('')
  return DOMPurify.sanitize(html, {
    ALLOWED_TAGS,
    ALLOWED_ATTR,
    ALLOW_DATA_ATTR: false,
  })
}
